adapters/zemanta/zemanta.go

```diff
@@ -0,0 +1,147 @@
+
+	var errs []error
+	var zemantaExt openrtb_ext.ExtImpZemanta
+	for i := 0; i < len(reqCopy.Imp); i++ {
+		imp := reqCopy.Imp[i]
+
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if err := json.Unmarshal(bidderExt.Bidder, &zemantaExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		imp.TagID = zemantaExt.TagId
+		reqCopy.Imp[i] = imp
+	}
+
+	publisher := &openrtb.Publisher{
+		ID:     zemantaExt.Publisher.Id,
+		Name:   zemantaExt.Publisher.Name,
+		Domain: zemantaExt.Publisher.Domain,
+	}
+	if reqCopy.Site != nil {
+		siteCopy := *reqCopy.Site
+		siteCopy.Publisher = publisher
+		reqCopy.Site = &siteCopy
+	} else if reqCopy.App != nil {
+		appCopy := *reqCopy.App
+		appCopy.Publisher = publisher
+		reqCopy.App = &appCopy
+	}
+
+	if zemantaExt.BCat != nil {
+		reqCopy.BCat = zemantaExt.BCat
+	}
+	if zemantaExt.BAdv != nil {
+		reqCopy.BAdv = zemantaExt.BAdv
+	}
+
+	requestJSON, err := json.Marshal(reqCopy)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+
+	var errs []error
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bidType, err := getMediaTypeForImp(bid.ImpID, request.Imp)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			b := &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, errs
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb.Imp) (openrtb_ext.BidType, error) {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil {
+				mediaType = openrtb_ext.BidTypeNative
+			}
+			return mediaType, nil
+		}
+	}
```

Comment:
This is an anti-pattern. Does your bidding server provide metadata in the response describing if the bid is a native or banner ad?

Reply:
Unfortunately no, we don't send any info about the media type in the response.

Reply:
Gotcha. Does this logic best mimic the behavior of your server? Will it prefer banner over native?

Reply:
Thanks for pointing this out, it actually wasn't. Fixed now.